qualtran/bloqs/generalized_qsp.py

```diff
@@ -79,6 +79,88 @@ def decompose_from_registers(
         yield cirq.GlobalPhaseGate(matrix[0, 0].conj()).on()
 
 
+def qsp_complementary_polynomial(
+    P: Sequence[complex], *, verify: bool = False
+) -> Sequence[complex]:
+    r"""Computes the Q polynomial given P
+
+    Computes polynomial $Q$ of degree at-most that of $P$, satisfying
+
+        $$ \abs{P(e^{i\theta})}^2 + \abs{Q(e^{i\theta})}^2 = 1 $$
+
+    for every $\theta \in \mathbb{R}$.
+
+    The exact method for computing $Q$ is described in the proof of Theorem 4.
+
+    Args:
+        P: Co-efficients of a complex polynomial.
+        verify: sanity check the computed polynomial roots (defaults to False).
+
+    References:
+        [Generalized Quantum Signal Processing](https://arxiv.org/abs/2308.01501)
+            Motlagh and Wiebe. (2023). Theorem 4.
+    """
+    d = len(P) - 1  # degree
+
+    # R(z) = z^d (1 - P^*(z) P(z)) (Eq. 34, 35)
+    R = Polynomial.basis(d) - Polynomial(P) * Polynomial(np.conj(P[::-1]))
+    roots = R.roots()
+
+    units: list[complex] = []  # roots r s.t. \abs{r} = 1
+    larger_roots: list[complex] = []  # roots r s.t. \abs{r} > 1
+    smaller_roots: list[complex] = []  # roots r s.t. \abs{r} < 1
+
+    for r in roots:
```

Comment:
It'd be great if you can add more comments throughout the implementation, mapping it back to the proof of Theorem 4 since I think there are a few differences in the explanation and your implementation? More comments to make it easier for readers to follow

Reply:
I added comments inline explaining the mapping and respective eqs/paras in the paper. Could you see if something is missing?